### PR TITLE
Super Mega Mob Spawn Refactor

### PR DESCRIPTION
**Pull request by tralezab**
## About The Super Hyper Ultra Ultimate Deluxe Perfect Amazing Shining Mob Spawn Refactor

The Super Hyper Ultra Ultimate Deluxe Perfect Amazing Shining Mob Spawn Refactor is my attempt to clean up the file structure, the code, and the type tree for mob spawns.

* Splits mob spawn types into corpses (dead spawns) and ghost roles (living spawns you can possess). The vars that didn't make sense for corpses and vice versa for ghost roles are now appropriately there
* Because of above, there are no longer the fucking "death, roundstart, and instant" vars. thank god
* Removes a lot of single or very few used vars, whose properties can be applied on special().
* All Mob Spawns are given fitting folders instead of just being stuck in a single ghost roles file. Corpses are in the corpse folder, Ghost Roles are in the ghost role folder. Only exception are drones which should stay near their respective homes
* Just generally cleaner all around you know
* spider structures file renamed to spiderwebs now that spider eggs are gone

## Why Super Hyper Ultra Ultimate Deluxe Perfect Amazing Shining Mob Spawn Refactor Is Good For The Game

The Super Hyper Ultra Ultimate Deluxe Perfect Amazing Shining Mob Spawn Refactor cleans up so many terrible cases and uses

## Changelog For The Super Hyper Ultra Ultimate Deluxe Perfect Amazing Shining Mob Spawn Refactor

:cl: armhulen
refactor: Mob spawns are refactored, no more assortment of "random, instant, and roundstart" vars on every mob spawn type
refactor: if there are some minimal differences in how mob spawners feel, that's why!
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
Oh I need to do the unit test as well, one moment

File: code/modules/mob_spawn/corpses/mining_corpses.dm [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+/datum/outfit/consumed_clown/pre_equip(mob/living/carbon/human/clown, visualsOnly = FALSE)
+	if(!visualsOnly)
+		clown.fully_replace_character_name(clown.name, pick(GLOB.clown_names))
+	if(prob(70))
+		backpack_contents += pick(list(/obj/item/stamp/clown = 1, /obj/item/reagent_containers/spray/waterflower = 1, /obj/item/food/grown/banana = 1, /obj/item/megaphone/clown = 1, /obj/item/reagent_containers/food/drinks/soda_cans/canned_laughter = 1, /obj/item/pneumatic_cannon/pie = 1))
+	if(prob(30))
+		backpack_contents += list(/obj/item/stack/sheet/mineral/bananium = pick_weight(list( 1 = 3, 2 = 2, 3 = 1)))
+	if(prob(10))
+		l_pocket = pick_weight(list(/obj/item/bikehorn/golden = 3, /obj/item/bikehorn/airhorn= 1 ))
```
Comment (SuperNovaa41) [inline code/modules/mob_spawn/corpses/mining_corpses.dm:144]:
```suggestion
		l_pocket = pick_weight(list(/obj/item/bikehorn/golden = 3, /obj/item/bikehorn/airhorn = 1 ))
```
:anger:

File: code/modules/mob_spawn/ghost_roles/golem_roles.dm [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	icon = 'icons/obj/wizard.dmi'
+	icon_state = "construct"
+	mob_species = /datum/species/golem
+	anchored = FALSE
+	move_resist = MOVE_FORCE_NORMAL
+	density = FALSE
+	var/has_owner = FALSE
+	var/can_transfer = TRUE //if golems can switch bodies to this new shell
+	var/mob/living/owner = null //golem's owner if it has one
+	you_are_text = "You are a Free Golem. Your family worships The Liberator."
```
Comment (SuperNovaa41) [inline code/modules/mob_spawn/ghost_roles/golem_roles.dm:18]: these should be before the new var declarations

File: code/modules/mob_spawn/ghost_roles/golem_roles.dm [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	flavour_text = "In his infinite and divine wisdom, he set your clan free to travel the stars with a single declaration: \"Yeah go do whatever.\""
+
+/obj/effect/mob_spawn/ghost_role/human/golem/Initialize(mapload, datum/species/golem/species = null, mob/creator = null)
+	if(species) //spawners list uses object name to register so this goes before ..()
+		name += " ([initial(species.prefix)])"
+		mob_species = species
+	. = ..()
+	var/area/A = get_area(src)
```
Comment (SuperNovaa41) [inline code/modules/mob_spawn/ghost_roles/golem_roles.dm:26]: :angry: (i can't tell if you added this or moved this so i won't mention the numerous others in this file)

File: code/modules/mob_spawn/ghost_roles/mining_roles.dm [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+/datum/outfit/beachbum
+	name = "Beach Bum"
+	glasses = /obj/item/clothing/glasses/sunglasses
+	r_pocket = /obj/item/storage/wallet/random
+	l_pocket = /obj/item/food/pizzaslice/dank
+	uniform = /obj/item/clothing/under/pants/youngfolksjeans
+	id = /obj/item/card/id/advanced
+
+/datum/outfit/beachbum/post_equip(mob/living/carbon/human/H, visualsOnly = FALSE)
```
Comment (SuperNovaa41) [inline code/modules/mob_spawn/ghost_roles/mining_roles.dm:100]: same here

**Review by SuperNovaa41 — COMMENTED** [review]
i can't tell for most of these if they're from you or if you're just moving around so i will give you the benefit of the doubt for all the shitty var names

File: code/modules/mob_spawn/corpses/mining_corpses.dm [hunk 5]
```diff
@@ -207,5 +207,5 @@
 	suit_store = /obj/item/tome
 	l_pocket = /obj/item/melee/cultblade/dagger
 	back = /obj/item/storage/backpack/cultpack
-	glasses = /obj/item/clothing/glasses/hud/health/night/cultblind
+	r_pocket = /obj/item/clothing/glasses/hud/health/night/cultblind
```
Comment (Mothblocks) [inline code/modules/mob_spawn/corpses/mining_corpses.dm:210]: wut

File: _maps/map_files/CTF/cruiser.dmm [hunk 6]
```diff
@@ -253,7 +253,7 @@
 /turf/open/floor/iron,
 /area/ctf)
 "mb" = (
-/obj/effect/mob_spawn/human/abductor,
+/obj/effect/mob_spawn/corpse/human/abductor,
```
Comment (NamelessFairy) [inline _maps/map_files/CTF/cruiser.dmm:256]: this one line change to a CTF map seems fine 👍

File: code/modules/mob_spawn/corpses/mining_corpses.dm (hunk 5 above)
Comment (tralezab) [reply code/modules/mob_spawn/corpses/mining_corpses.dm:210]: cult blind glasses literally can not be equipped on corpses along with a lot of other gear. the mob spawn was originally dodging the unit tests by overriding outfit slots

File: _maps/map_files/CTF/cruiser.dmm (hunk 6 above)
Comment (tralezab) [reply _maps/map_files/CTF/cruiser.dmm:256]: kek

File: code/modules/mob_spawn/corpses/mining_corpses.dm (hunk 5 above)
Comment (Mothblocks) [reply code/modules/mob_spawn/corpses/mining_corpses.dm:210]: haha

**Review by Fikou — APPROVED** [review]
lgtm

File: code/modules/mob_spawn/mob_spawn.dm [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+	///Lazy string list of factions that the spawned mob will be in upon spawn
+	var/list/faction
+
+	////Human specific stuff. Don't set these if you aren't using a human, the unit tests will put a stop to your sinful hand.
+
+	///sets the human as a species, use a typepath (example: /datum/species/skeleton)
+	var/mob_species
+	///equips the human with an outfit.
+	var/datum/outfit/outfit
+	///for mappers to override parts of the outfit. really only in here for secret away missions, please try to refrain from using this out of laziness
```
Comment (Fikou) [inline code/modules/mob_spawn/mob_spawn.dm:21]: you use this on 2 ruins so its not only secret away missions

File: code/modules/mob_spawn/mob_spawn.dm [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+	else if(mob_name)
+		chosen_name = mob_name
+	//then if no name was chosen the one the mob has by default works great
+	if(!chosen_name)
+		return
+	//not using an old name doesn't update records- but ghost roles don't have records so who cares
+	spawned_mob.fully_replace_character_name(null, chosen_name)
+
+/obj/effect/mob_spawn/proc/equip(mob/living/spawned_mob)
+	if(outfit)
```
Comment (Fikou) [inline code/modules/mob_spawn/mob_spawn.dm:95]: early return?

File: code/modules/mob_spawn/mob_spawn.dm [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+/obj/effect/mob_spawn/corpse/human/special(mob/living/carbon/human/spawned_human)
+	. = ..()
+	if(husk)
+		spawned_human.Drain()
+	else //Because for some reason I can't track down, things are getting turned into husks even if husk = false. It's in some damage proc somewhere.
+		spawned_human.cure_husk()
+
+/obj/effect/mob_spawn/corpse/human/equip(mob/living/carbon/human/spawned_human)
+	. = ..()
+	if(conceal_presence)
```
Comment (Fikou) [inline code/modules/mob_spawn/mob_spawn.dm:254]: early return

**Pull request merged**